### ISSUE TITLE
About dataloader for multi-source dataset and training setting

### ISSUE THREAD

**Issue by sglucas** [ZHZisZZ/dllm-trainer#83]
Hi, thank you for your great work.

I’m planning to follow the Dream training pipeline to pre-train a new model with `dllm` framework. I first convert an AR model to a DLLM model, and then use the`dllm` setup to start pre-training. For the pre-training process, I’m following `examples/a2d/mdlm/pt.py`, and I have a few questions:

1. Does the current dataloader support multi-source datasets? Ideally, I’d like each batch to sample examples from different sources (mixture sampling).

2. For MDLM evaluation, which script or entry point should I follow? I noticed --mode is defined differently across these options:

`bash examples/a2d/bd3lm/eval.sh`

`bash examples/a2d/mdlm/eval.sh`

`accelerate launch --num_processes 4 \
    dllm/pipelines/a2d/eval.py \
    --tasks "gsm8k_cot" \
    --model "mdlm" \
    --apply_chat_template \
    --num_fewshot 0 \
    --model_args "pretrained=dllm-collection/Qwen3-0.6B-diffusion-mdlm-v0.1,max_new_tokens=256,steps=256,block_size=256,cfg=0.0,temperature=0.0"`
Which one is the recommended/equivalent way to evaluate MDLM, and how should --mode be set?

3. I tried training for ~4B tokens, and the loss seems to decrease, then increase, then decrease again, repeatedly. I’m not sure if this behavior is expected for MDLM pre-training, or if it indicates a data/config issue.

```
9.9625
9.4304
6.7585
5.2066
4.3318
3.9792
3.4891
3.3677
**3.2183**
**3.8668**
3.8697
3.7018
3.6339
3.3171
3.2667
3.1283
**3.0035**
**3.3031**
**3.6949**
3.7335
3.474
3.1321
3.0166
**2.9727**
**3.022**
**3.5955**
3.6327
3.5994
3.1921
**3.0323**
**3.1235**
**3.5689**
3.6257
```

Thank you very much for your help in advance!

**Comment by ZHZisZZ** [comment]
1. You can try something like this: https://github.com/ZHZisZZ/dllm/blob/main/dllm/data/utils.py#L25
2. You should use `bash examples/a2d/mdlm/eval.sh --model_name_or_path "dllm-collection/Qwen3-0.6B-diffusion-mdlm-v0.1"`; Sorry for the typo in the README.md. 
3. This is completely normal. Since different tokens are masked every epoch, you should expect these fluctuations during LLM training.

We are refactoring the evaluation pipeline to remove redundant code; however, the current scripts remain functional.

**Comment by sglucas** [comment]
Thank you very much for your response! May I ask whether you try to evaluate the model with this training command:
```
accelerate launch --config_file scripts/accelerate_configs/zero2.yaml --num_processes 8 \
    examples/a2d/mdlm/pt.py \
    --model_name_or_path "models/a2d/Qwen3-0.6B" \
    --dataset_args "dylanebert/openwebtext" \
    --text_field "text" \
    --streaming True \
    --insert_eos True \
    --max_length 512 \
    --max_steps 20000 \
    --learning_rate 1e-4 \
    --per_device_train_batch_size 16 \
    --per_device_eval_batch_size 16 \
    --eval_strategy "no" \
    --output_dir "models/a2d/Qwen3-0.6B/mdlm/openwebtext"
```
I believe this command is suitable for model pre-training and also try to follow this training command to obtain a pre-trained model. When I try to evaluate the performance on mmlu, gsm8k, I found the performance is not very good. So I would like to ask whether you evaluate the model before to check whether my current setting is correct.

**Comment by sglucas** [comment]
Hi, I would like to update: when I increase the max_length from 500 to 4096, I found the gradient norm become very large: like 86445441024.0, although we use gradient clip. I'm not sure whether there is something wrong.

**Comment by qky18** [comment]
> Hi, I would like to update: when I increase the max_length from 500 to 4096, I found the gradient norm become very large: like 86445441024.0, although we use gradient clip. I'm not sure whether there is something wrong.
> 
> In addition, I would like to ask whether I need to change `time_epsilon` when I use a large max_length, like 4096.

Hi I've also encountered the same issue, and during debugging I traced this back to some memory spike in backward propagation that created garbage gradients but not enough to raise an OOM. I solved this by restricting the loss computation to masked tokens only via indexing to save memory, rather than by multiplicating a binary mask as originally implemented.

Here's my code in trainer's function `forward` for reference:
```python
        # === 5. Memory-efficient cross-entropy (only at masked positions) ===
        # Computing CE on the full [b, l, V] tensor materializes a float32 softmax
        # of ~17 GB for large vocab (V~150k) and batch sizes, causing OOM during
        # backward. Instead, gather only masked positions → [num_masked, V].
        assert (
            input_ids[maskable_mask] == labels[maskable_mask]
        ).all(), "Mismatch between input_ids and labels at valid positions"

        masked_indices = masked_mask.nonzero(as_tuple=True)  # (batch_idx, pos_idx)
        masked_logits = logits[masked_indices[0], masked_indices[1]]  # [num_masked, V]
        masked_targets = input_ids[masked_indices[0], masked_indices[1]]  # [num_masked]
        masked_nll = F.cross_entropy(masked_logits, masked_targets, reduction="none")

        # Apply per-sample loss weights at masked positions
        masked_loss_weights = loss_weights[masked_indices[0], masked_indices[1]]
        masked_nll = masked_nll * masked_loss_weights  # [num_masked]
```
